Update to simplified roles structure for Public

oyving opened the pull request:
Create the new roles structure for Public - or at least I tried to. These changes probably break existing applications in Public, so we might have to re-introduce some things?

Don't merge, I think we should discuss the names of the constants as well.

Commit:
Update to simplified roles structure for Public

jonmv commented on controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/user/Roles.java:
```diff
@@ -45,20 +44,17 @@ public static Role toRole(String value) {
-            case "tenantOperator": return Role.tenantOperator(tenant);
+            case "publicReader": return Role.publicReader(tenant);
+            case "publicUser": return Role.publicDeveloper(tenant);
+            case "publicAdministrator": return Role.publicAdministrator(tenant);
```
Changing the names of the roles here means all existing role memberships are lost. This is probably OK.

oyving replied:
Yes, I talked to @smorgrav about this, and we thought it would be ok. _However_ I am not sure if I like the names :)

jonmv commented on controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java:
```diff
@@ -20,60 +20,46 @@
+    /** Headless - the application specific role identified by deployment keys for production */
+    publicHeadless(
+            Policy.submission,
+            Policy.deploymentPipeline,
```
Legacy.

oyving replied:
So `Policy.deploymentPipeline` can be removed?

jonmv replied:
Yes.

jonmv commented on controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java:
```diff
@@ -20,60 +20,46 @@
+    publicHeadless(
+            Policy.submission,
+            Policy.deploymentPipeline,
+            Policy.productionDeployment
```
Legacy.

oyving replied:
Ditto on removal?

jonmv replied:
Ditto.

jonmv commented on controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java:
```diff
@@ -20,60 +20,46 @@
+    ),
+
+    /** Headless - the application specific role identified by deployment keys for production */
+    publicHeadless(
```
I think the `public` prefix could be dropped from all these names. 

The `publicRead` privilege refers to the common word "public", not the system(s) :)

oyving replied:
As I mentioned before - I am not a huge fan of the names in the enum, so I am all for removing the prefix.

jonmv replied:
Great!

jonmv commented on controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java:
```diff
@@ -20,60 +20,46 @@
+            Policy.deploymentPipeline,
+            Policy.productionDeployment
+    ),
+
```
A role which a `Developer` assumes when authenticating through a signed request is missing. I believe we wanted to include only the "read" and "development deployment" policies in that.

oyving replied:
Nice catch. `developerWithKey` or something like that?

jonmv replied:
I suppose.

jonmv commented on controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java:
```diff
@@ -20,60 +20,46 @@
+    /** Admin - the administrative function for user management etc. */
+    publicAdministrator(
+            Policy.tenantUpdate,
+            Policy.tenantManager
```
Only we may delete tenants?

jonmv replied:
Missing the `applicationManager` policy.

oyving replied:
Yes, I think it makes sense that only we delete tenants. I'll look into `applicationManager`.

jonmv replied:
We _could_ remove some granularity in the underlying policies and privileges, but I'm not sure we should, at least not before the new outwards model is set in stone.

oyving replied:
No, I don't think we should remove granularity there. I'd rather we bucket things together at the top :)

jonmv commented on controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java:
```diff
@@ -20,60 +20,46 @@
+            Policy.applicationCreate,
+            Policy.applicationUpdate,
+            Policy.applicationDelete,
+            Policy.developmentDeployment
```
Missing the `applicationOperations` policy.

jonmv commented on controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/filter/SignatureFilter.java:
```diff
@@ -60,9 +60,7 @@ public SignatureFilter(Controller controller) {
-                                         new SecurityContext(principal,
-                                                             Set.of(Role.buildService(id.tenant(), id.application()),
-                                                                    Role.applicationDeveloper(id.tenant(), id.application()))));
+                                         new SecurityContext(principal, Set.of(Role.publicHeadless(id.tenant(), id.application()))));
```
Can we hold this change until developer keys are in place?

jonmv commented on controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/user/UserApiHandler.java:
```diff
@@ -211,17 +211,4 @@ private static Inspector bodyInspector(HttpRequest request) {
-            case applicationOperator:   return "applicationOperator";
-            case applicationDeveloper:  return "applicationDeveloper";
-            case applicationReader:     return "applicationReader";
-            default: throw new IllegalArgumentException("Unexpected role type '" + role.definition() + "'.");
```
I would keep both these, as they are used in different serialisation contexts.

oyving replied:
Added it back.

jonmv reviewed (changes requested):
It's easier to deal with roles when there is no inheritance, which is feasible when there are fewer roles, so I think these changes are good. There are some comments to address.

oyving commented on controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/user/Roles.java:
```diff
@@ -20,17 +20,16 @@ private Roles() { }
-                       Role.tenantOperator(tenant));
+        return List.of(
+                Role.publicReader(tenant),
+                Role.publicDeveloper(tenant),
```
Correct. That was probably only implied in the design doc. that we circulated internally.

Commit:
Remove public* prefix

Commit:
Remove legacy privileges from headless

Commit:
Add applicationManager to administrator

Not sure if this makes sense as long as user management at the application level
is not something we intend to use.

Commit:
Make developers able to restart prod instances

Commit:
Keep local serializer

Commit:
Fix test values

jonmv commented:
Very nice!

mpolden commented:
Much better!

oyving commented:
Want to coordinate merge with UI work that happens this sprint.

jonmv commented on controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java:
```diff
@@ -20,60 +20,46 @@
+
+    /** Admin - the administrative function for user management etc. */
+    administrator(
+            Policy.tenantUpdate,
```
This should be moved to developer. It currently covers nothing, and will cover developer key management.

jonmv replied:
Hmm... We would probably want to ability to change tenant billing settings at some point, though. Perhaps we rather need two separate paths.

jonmv replied:
Ok, moved things around:

Deploy keys are now POST/DELETE at `application/v4/tenant/{tenant}/application/{application}/key`, and developer keys at `application/v4/tenant/{tenant}/key`.

There's a `Policy.keyManagement` to allow these actions. Please insert this under the `developer` role above.

jonmv replied:
(When it's available.)

jonmv replied:
Or, I rather put the key stuff under tenantOperator for now, just for testing. It was simpler.

Commit:
Merge remote-tracking branch 'origin/master' into ogronnesby/new-roles-for-public

jonmv commented on controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/user/Roles.java:
```diff
@@ -45,20 +44,17 @@ public static Role toRole(String value) {
-            case "tenantAdmin": return Role.tenantAdmin(tenant);
-            case "tenantOperator": return Role.tenantOperator(tenant);
+            case "reader": return Role.reader(tenant);
+            case "user": return Role.developer(tenant);
```
`user` or `developer`?

oyving replied:
`developer` :)

Commit:
Fix some remaining tokens from user -> developer renaming

Commit:
Use new roles for filter testing

Commit:
Update UserApi test for new roles

Commit:
Merge remote-tracking branch 'origin/master' into ogronnesby/new-roles-for-public

Conflicts:
	controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java
	controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/user/UserApiHandler.java
	controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/user/UserApiTest.java

Commit:
Give keyManagement to developers and update tests after merge

jonmv commented:
Oh, and if we're to allow submissions from the console, we also need the `submission` policy for `developer`s.

Commit:
Merge remote-tracking branch 'origin/master' into ogronnesby/new-roles-for-public

Conflicts:
	controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/Role.java
	controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/RoleDefinition.java
	controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/filter/SignatureFilter.java
	controller-server/src/main/java/com/yahoo/vespa/hosted/controller/security/CloudAccessControl.java
	controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/filter/SignatureFilterTest.java
	controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/user/UserApiTest.java

oyving commented:
Trying to remake this here: https://github.com/vespa-engine/vespa/pull/10921
